src/v/model/record.h

```diff
@@ -647,16 +670,32 @@ class record_batch {
     const iobuf& data() const { return _records; }
     iobuf&& release_data() && { return std::move(_records); }
     void clear_data() { _records.clear(); }
+    void key_offsets(std::vector<key_offset>&& key_offsets) {
+        _key_offsets = ss::make_lw_shared<std::vector<key_offset>>(
+          std::move(key_offsets));
+    }
+    const ss::lw_shared_ptr<std::vector<key_offset>> key_offsets() const {
+        return _key_offsets;
+    }
+    ss::lw_shared_ptr<std::vector<key_offset>> key_offsets() {
+        return _key_offsets;
+    }
 
 private:
     record_batch_header _header;
     iobuf _records;
     bool _compressed;
+    ss::lw_shared_ptr<std::vector<key_offset>> _key_offsets;
 
-    record_batch(record_batch_header header, iobuf&& records, bool compressed)
+    record_batch(
+      record_batch_header header,
+      iobuf&& records,
+      bool compressed,
+      ss::lw_shared_ptr<std::vector<key_offset>> key_offsets)
```

Comment:
it doesn't have to be shared pointer, `model::record_batch` is not copyable

Reply:
yes but we still copy it by calling `.share()` inside `src/v/raft/consensus_utils.cc` so if we make it a vector we'd need to deep copy instead of just copying a pointer

Reply:
I would say that it would be more safe to make a deep copy when it comes to cross core accesses. Also this is the thing to consider if we need to wrap `key_offsets` with `foreign_pointer`